flink-core/src/main/java/org/apache/flink/core/fs/SafetyNetCloseableRegistry.java

```diff
@@ -0,0 +1,165 @@
+ * <p>
+ * Phantom references are used to track when {@link org.apache.flink.util.WrappingProxy}s of {@link Closeable} got
+ * GC'ed. We ensure that the wrapped {@link Closeable} is proeprly closed to avoid resource leaks.
+ * <p>
+ * Other than that, it works like a normal {@link CloseableRegistry}.
+ * <p>
+ * All methods in this class are thread-safe.
+ */
+public class SafetyNetCloseableRegistry extends
+		AbstractCloseableRegistry<WrappingProxyCloseable<? extends Closeable>,
+				SafetyNetCloseableRegistry.PhantomDelegatingCloseableRef> {
+
+	private static final Logger LOG = LoggerFactory.getLogger(SafetyNetCloseableRegistry.class);
+	private final ReferenceQueue<WrappingProxyCloseable<? extends Closeable>> referenceQueue;
+	private final Thread reaperThread;
+
+	public SafetyNetCloseableRegistry() {
+		super(new IdentityHashMap<Closeable, PhantomDelegatingCloseableRef>());
+		this.referenceQueue = new ReferenceQueue<>();
+		this.reaperThread = new Thread(new CloseableReaperRunnable());
+		reaperThread.start();
+	}
+
+	@Override
+	protected void doRegister(
+			WrappingProxyCloseable<? extends Closeable> wrappingProxyCloseable,
+			Map<Closeable, PhantomDelegatingCloseableRef> closeableMap) throws IOException {
+
+		Closeable innerCloseable = WrappingProxyUtil.stripProxy(wrappingProxyCloseable.getWrappedDelegate());
+
+		if (null == innerCloseable) {
+			return;
+		}
+
+		PhantomDelegatingCloseableRef phantomRef =
+				new PhantomDelegatingCloseableRef(wrappingProxyCloseable, referenceQueue);
+
+		closeableMap.put(innerCloseable, phantomRef);
+	}
+
+	@Override
+	protected void doUnRegister(
+			WrappingProxyCloseable<? extends Closeable> closeable,
+			Map<Closeable, PhantomDelegatingCloseableRef> closeableMap) {
+
+		Closeable innerCloseable = WrappingProxyUtil.stripProxy(closeable.getWrappedDelegate());
+
+		if (null == innerCloseable) {
+			return;
+		}
+
+		closeableMap.remove(innerCloseable);
+	}
+
+	/**
+	 * Phantom reference to {@link WrappingProxyCloseable}.
+	 */
+	static final class PhantomDelegatingCloseableRef
+			extends PhantomReference<WrappingProxyCloseable<? extends Closeable>>
+			implements Closeable {
+
+		private final Closeable innerCloseable;
+		private final String debugString;
+
+		public PhantomDelegatingCloseableRef(
+				WrappingProxyCloseable<? extends Closeable> referent,
+				ReferenceQueue<? super WrappingProxyCloseable<? extends Closeable>> q) {
+
+			super(referent, q);
+			this.innerCloseable = Preconditions.checkNotNull(WrappingProxyUtil.stripProxy(referent));
+			this.debugString = referent.toString();
+		}
+
+		public Closeable getInnerCloseable() {
+			return innerCloseable;
+		}
+
+		public String getDebugString() {
+			return debugString;
+		}
+
+		@Override
+		public void close() throws IOException {
+			innerCloseable.close();
+		}
+	}
+
+	/**
+	 * Reaper runnable collects and closes leaking resources
+	 */
+	final class CloseableReaperRunnable implements Runnable {
```

Comment:
I am usually against terminating threads purely via Interrupts. Its quite dangerous, because it only ever shows when the thread enters a blocking operation.
It will probably work here (not too much to do and frequently entering blocking calls), but to keep the code in the shape that it sets a good example for other contributors, lets add a proper `running` flag to control the `while` loop.